libknet/transport_common.c

```diff
@@ -87,31 +88,31 @@ static int _configure_sockbuf (knet_handle_t knet_h, int sock, int option, int f
 		if (value_len == sizeof new_value && target <= new_value) {
 			return 0;
 		}
-		errno = ERANGE;
-	}
-
-	log_debug (knet_h, KNET_SUB_TRANSPORT, "Failed to set socket buffer via option %d to value %d: %s",
-		   option, target, strerror(errno));
+		if (!force) {
+			errno = ERANGE;
+			log_debug (knet_h, KNET_SUB_TRANSPORT, "Failed to set socket buffer via option %d to value %d: %s",
+				   option, target, strerror(errno));
 
-	if (force) {
-		if (setsockopt(sock, SOL_SOCKET, force, &target, sizeof target)) {
-			int savederrno = errno;
-			log_debug (knet_h, KNET_SUB_TRANSPORT,
-				   "Failed to set socket buffer via fallback option %d as well: %s",
-				   force, strerror(errno));
-			errno = savederrno;
 			return -1;
-		} else {
-			return 0;
 		}
-	} else {
+	}
+
+	if (setsockopt(sock, SOL_SOCKET, force, &target, sizeof target) < 0) {
+		savederrno = errno;
+		log_debug (knet_h, KNET_SUB_TRANSPORT,
+			   "Failed to set socket buffer via force option %d: %s",
+			   force, strerror(errno));
+		errno = savederrno;
 		return -1;
 	}
+
+	return 0;
 }
 
 int _configure_common_socket(knet_handle_t knet_h, int sock, uint64_t flags, const char *type)
 {
 	int err = 0, savederrno = 0;
+	int value;
```

Comment:
Annoyingly enough, this makes the BSD builds fail with unused variable.

Reply:
Yes indeed. the issue is a missing include from transport_common.c :-)

Reply:
You certainly know your stuff: found a hidden bug behind a careless one! 👍
If you know that at least one of those conditionals will always be true, then moving the definition to the front won't cause problems. But why is it preferable? A trap for possible other hidden bugs? :)

Reply:
In this specific case, they were both supposed to be true (either in Linux or BSD), but clearly we need to do a better job at noticing those build differences. I think we need to either start printing more stuff at configure time or like the last commit, at runtime.